Rework some parts of the card database

ctrlaltca opened the pull request:
This PR aims to improve the way the card database works in order to simplify it and make it more easy to work with it.

The first notable change is that now card database loading is progressive, and runs in another thread. Cockatrice starts up a lot faster, and the deck editor populates the card list in the background.

The second notable change is that when cockatrice is not able to load the card database, it won't just show the settings page. Instead, it will propose the user to run oracle and download a fresh copy of the card database:

![welcome](https://cloud.githubusercontent.com/assets/1631111/13732153/da0ff20e-e97b-11e5-8eb4-f70dea59dcde.png)

As a result, the NSIS installer now proposes to run Cockatrice instead of to run Oracle. The first-user workflow is: installer → cockatrice → download db dialog → run oracle → ready to play

The third notable change is that some actions related to the card database (like edit sets, edit tokens) have been moved from the deckeditor tab class to the mainwindow class, to permit to access those actions even if no deck editor is open.

The fourth notable change is that now Cockarice saves user-created tokens on a different file, to avoid them being lost when oracle is run.

Minor changes:
 * in the deck editor, the "name" column now fills up all the available space;
 * PictureLoader now avoids trying to download twice the same card picture at the same time;
 * The “edit tokens” dialog now actually works;

More technicals changes:
 * the card database loading and checking methods have been unified in a single method inside the CardDatabase class;
 * all the directory checking calls in main.cpp have been moved to the settingscache, so that any path returned by settingscache is supposed to be safe (existing).
 * the database reload after "edit sets" has been optimized to be faster;
 * CardDatabase: removed the noCard hack.
 * CardDatabase::getCard() is now const, since it no more creates cards instead of just returning existing
ones
 * AbstractCardItem: avoid recalculating card color at every paint

Fix #1858, fix #1768, fix #1531, fix #1861, fix #1843, fix #1546, fix #1896

Daenyth commented on cockatrice/CMakeLists.txt:
```diff
@@ -175,6 +175,13 @@ if(Qt5Widgets_FOUND)
 
+    # QtConcurrent
+    find_package(Qt5Concurrent)
+    if(Qt5Concurrent_FOUND)
```
Do we need any new ubuntu/brew installs? Windows setup?

ctrlaltca replied:
No, it's already used in oracle.

Daenyth commented on cockatrice/src/abstractcarditem.cpp:
```diff
@@ -90,44 +93,21 @@ void AbstractCardItem::paintPicture(QPainter *painter, const QSizeF &translatedS
     qreal scaleFactor = translatedSize.width() / boundingRect().width();
     
-    CardInfo *imageSource = facedown ? db->getCard() : info;
+    CardInfo *imageSource = facedown ? 0 : info;
```
Maybe `nullptr` instead of `0` here?

ctrlaltca replied:
Good idea, c++11 ftw

Daenyth commented on cockatrice/src/carddatabase.cpp:
```diff
@@ -341,39 +341,35 @@ static QXmlStreamWriter &operator<<(QXmlStreamWriter &xml, const CardInfo *info)
+    QHashIterator<QString, CardSet *> setIt(sets);
+    while (setIt.hasNext()) {
+        setIt.next();
+        delete setIt.value();
```
Not `deleteLater`?

ctrlaltca replied:
CardSet is a QList; a QList doesn't inherit from QObject, so there's no deleteLater() method available.

Daenyth replied:
:+1:

Daenyth commented on cockatrice/src/carddatabase.cpp:
```diff
@@ -518,28 +515,20 @@ void CardDatabase::loadCardsFromXml(QXmlStreamReader &xml, bool tokens)
-    }
 
-    return noCard;
+    return 0;
```
`nullptr`

Daenyth commented on cockatrice/src/carddatabase.cpp:
```diff
@@ -562,11 +551,10 @@ LoadStatus CardDatabase::loadFromFile(const QString &fileName, bool tokens)
             }
         }
     }
-    qDebug() << cards.size() << "cards in" << sets.size() << "sets loaded";
```
Is this moved elsewhere? It's useful for debugging

ctrlaltca replied:
Yes, it's been merged with another similar debug line at https://github.com/Cockatrice/Cockatrice/pull/1859/files#diff-753bd6437004806e213fe65d32cfa732R606

Daenyth commented on cockatrice/src/carddatabase.cpp:
```diff
@@ -609,57 +597,59 @@ bool CardDatabase::saveToFile(const QString &fileName, bool tokens)
+    // load main card database
+    loadStatus = loadCardDatabase(settingsCache->getCardDatabasePath());
+    // laod tokens database
+    loadCardDatabase(settingsCache->getTokenDatabasePath());
```
No loadStatus for this?

ctrlaltca replied:
By now this mimics the old behavior where only a main card database loading failure was considered a "blocking" problem, while it was fine tu run without a tokens database.

Daenyth replied:
I agree we should preserve that but it would be good to qDebug for linux/osx. Feel free to ticket it for later and ignore for now.

ctrlaltca replied:
Done: https://github.com/Cockatrice/Cockatrice/issues/1867

Daenyth commented on cockatrice/src/carddatabase.cpp:
```diff
@@ -609,57 +597,59 @@ bool CardDatabase::saveToFile(const QString &fileName, bool tokens)
+    QDir dir(settingsCache->getCustomCardDatabasePath());
+    foreach(QString fileName, dir.entryList(QStringList("*.xml"), QDir::Files | QDir::Readable, QDir::Name | QDir::IgnoreCase))
+    {
+        loadCardDatabase(dir.absoluteFilePath(fileName));
```
We should add error handling and logging to this to help people debug bad custom sets

ctrlaltca replied:
I agree, this would be really helpful.

Daenyth commented on cockatrice/src/abstractcarditem.cpp:
```diff
@@ -192,11 +172,11 @@ void AbstractCardItem::setName(const QString &_name)
     emit deleteCardInfoPopup(name);
-    disconnect(info, 0, this, 0);
+    if(info)
+        disconnect(info, 0, this, 0);
```
Should these also be nullptr?

ctrlaltca replied:
They could.

Daenyth commented on cockatrice/src/cardinfowidget.cpp:
```diff
@@ -183,11 +195,7 @@ void CardInfoWidget::setCard(AbstractCardItem *card)
-
-CardInfo *CardInfoWidget::getCard(const QString &cardName) {
-    return db->getCardBySimpleName(cardName);
+    setCard((CardInfo *) 0);
```
nullptr

Daenyth commented on cockatrice/src/cardinfowidget.cpp:
```diff
@@ -153,17 +153,29 @@ void CardInfoWidget::setCard(CardInfo *card)
+        cardtypeLabel2->setText("");
+        powtoughLabel2->setText("");
+        loyaltyLabel2->setText("");
+        textLabel->setText("");
```
Seems like these labels are somewhat duplicated from CardInfoText; maybe there should be some shared behavior between them to reduce it.

ctrlaltca replied:
:+1:

Daenyth commented on cockatrice/src/dlg_create_token.cpp:
```diff
@@ -132,14 +132,22 @@ DlgCreateToken::DlgCreateToken(const QStringList &_predefinedTokens, QWidget *pa
+    if(cardInfo)
+    {
+        nameEdit->setText(cardInfo->getName());
+        const QString cardColor = cardInfo->getColors().isEmpty() ? QString() : (cardInfo->getColors().size() > 1 ? QString("m") : cardInfo->getColors().first());
```
This could be a method on CardInfo

ctrlaltca replied:
:+1:

Daenyth commented on cockatrice/src/pictureloader.cpp:
```diff
@@ -428,30 +431,28 @@ void PictureLoader::getPixmap(QPixmap &pixmap, CardInfo *card, QSize size)
         // add the card to the load queue
         getInstance().worker->enqueueImageLoad(card);
+    } else {
+        // requesting the image for a null card is a shortcut to get the card background image
```
Good comment

Daenyth commented on cockatrice/src/settingscache.cpp:
```diff
@@ -9,20 +11,21 @@
-    return file;
+QString SettingsCache::getSettingsPath()
+{
+    return getDataPath() + "/settings/";
```
:+1:

Daenyth commented:
Overall this looks great; I'd like to give it very heavy testing, maybe publish out some beta builds to share in the chat

tooomm commented:
AWESOME! :+1: :+1: 

Please check if this also fixes https://github.com/Cockatrice/Cockatrice/issues/1531#issuecomment-176433386 and edit the start post!

I did some tests already and will write a roundup soon. Run into some stuff... (no worries :P)

tooomm commented:
This is going to be a longer post while going trough all your points... please read it carefully!. :grinning: 

<br>
>The first notable change is that now card database loading is progressive, and runs in another thread. Cockatrice starts up a lot faster, and the deck editor populates the card list in the background.

No delay on app start up anymore! From 3s delay down to zero mostly...
I highly prefer that (and don't mind that the card database refreshes/loads visibly in the background while the app is already useable!)

<br>
>The second notable change is that when cockatrice is not able to load the card database, it won't just show the settings page. Instead, it will propose the user to run oracle and download a fresh copy of the card database:

:+1: on the concept. (I just dislike the 2-3s delay until the dialog actually appears!)

We should also offer to import custom sets. (some people just play custom or other games than mtg)
Oracle also doesn't allow them to choose local xml file to create the database from, so that's no help either for them and they need to use the import feature since their database might not be named cards.xml.

<br>
>As a result, the NSIS installer now proposes to run Cockatrice instead of to run Oracle. The first-user workflow is: installer → cockatrice → download db dialog → run oracle → ready to play

This is great! Way better than before - users most likely won't recognize it. (and therefore love it!)

**It's not uniform yet, though** (I hope this is understandable and you can reproduce it, did so much deleting of cards.xml's, carddatabase.ini's, oracle run's and stuff :grinning: )
- first-run workflow (fresh first ever install, as you stated above):
installer → cockatrice → download db dialog → run oracle → ready to play :heavy_check_mark: 

- first-run workflow (after client update with **configs etc. NOT deleted**): 
installer → cockatrice → ready to play :heavy_check_mark:

- first-run workflow (after client update with **configs etc. deleted**):
installer → cockatrice → download db dialog → run oracle → **"import done, refresh database now" msg in cockarice without restart needed, hit `ok` → download db dialog** (you're in a loop here) :x:
If you hit `No` on the "download db dialog" you'll have a open and running Cockatrice with no cards, but after a few *long* seconds the old "Welcome, first run" msg appears and explains you that you need to manage your sets first (so it somehow still detects that you now have some cards, even the reloading didn't work!), you can just hit `ok` (actually hitting `x` will act the same as hitting `ok` - so you can't avoid it which is good and right here) and the `edit sets` window opens with all sets enabled by default. If you hit `ok` there Cockatrice starts loading the database, if you hit `x` or `cancel` Cockatrice shows up no cards, if you restart Cockatrice though all sets are enabled and show up - also a nice default if someone just clicks trough.<br>
**This needs some tweaks!** (checks are mistimed? refresh is triggered but leads to a repeating "download db" msg instead of a reload and the "welcome, first run" msg to appear...)

- second-use workflow:
cockatrice → ready to play :heavy_check_mark:

- new-set-release workflow (update card db button in app):
cockatrice → hit "check for card updates" → run oracle → "update finished, reloading database" msg **(huge lag here, Cockatrice mostly freezes for ~10s)** → ready to play (:heavy_check_mark:)<br>
**This needs a check!** (for the first test i just had a few sets enabled and it worked flawlessly, with all sets enabled before running the card update check I experienced this freeze)

- new-set-release workflow (new extended cards.xml file):
replace cards.xml file → cockatrice → ready to play :heavy_check_mark:

- new-set-release workflow (custom setx.xml file added):
*see "custom-set-added workflow" below*

- custom-set-added workflow:
cockatrice → hit "add custom sets/cards" → **"import of your newset.xml was successful, you need to restart" msg, hitting `ok` → new cards/sets don't show up until you restart manually**  :x:<br>
**This needs some tweaks!**
It should show the other "import done, cockatrice now reloads your db" msg from above instead + reloads the db after hitting `ok` instead of forcing the user to do a manual restart

<br>
- these are all I could think of and test right now... any misses?

<br>
>The third notable change is that some actions related to the card database (like edit sets, edit tokens) have been moved from the deckeditor tab class to the mainwindow class, to permit to access those actions even if no deck editor is open.

Very nice :+1: 
If you just play with downloaded or old decks or just do limited etc. (or you don't need the deckeditor for any reason), you could now still edit tokens or enable/disable sets and reorder them for picture loading purposes.

<br>
>The fourth notable change is that now Cockarice saves user-created tokens on a different file, to avoid them being lost when oracle is run.

I dont like the TK.xml filename. It doesn't provide any information, is confusing and might get mixed up with some set code or sth.
Go with `custom_token.xml` instead?

<br>
>in the deck editor, the "name" column now fills up all the available space;

:+1: 

<br>
>The “edit tokens” dialog now actually works

What didn't work before? I really can't tell, didn't use it in a while. :)

<br>
>CardDatabase: removed the noCard hack.

??

<br>
(more later)

Daenyth commented:
> > CardDatabase: removed the noCard hack.

> ??

There used to be a card named `noCard` as I recall for the card back. I might be misremembering

ctrlaltca commented:
> There used to be a card named noCard as I recall for the card back

Exactly. When the card database was requested a non-existing or empty-named card, it always returned a dummy card. This broke a lot of small things around, eg. the "does a card named xxx exist?" check.

ctrlaltca commented:
@tooomm thank you a lot for your deep testing, i'll review your feedback ASAP.

tooomm commented:
Take your time!

Since you address #1768, will this also fix https://github.com/Cockatrice/Cockatrice/issues/1861?

ctrlaltca commented:
> custom-set-added workflow:
> cockatrice → hit "add custom sets/cards" → "import of your newset.xml was successful, you need to restart" msg, hitting ok → new cards/sets don't show up until you restart manually  
> 
> This needs some tweaks!
> It should show the other "import done, cockatrice now reloads your db" msg from above instead + reloads the db after hitting ok instead of forcing the user to do a manual restart

Fixed

> Since you address #1768, will this also fix #1861?

Correct, added to PR description.

ctrlaltca commented:
> first-run workflow (after client update with configs etc. deleted):
> installer → cockatrice → download db dialog → run oracle → "import done, refresh database now" msg in cockarice without restart needed, hit ok → download db dialog (you're in a loop here) 

Somehow the database failed to load on this second attempt (but we don't know why, #1867). Maybe oracle didn't complete the card import successfully? If you can reproduce this, please check if cards.xml is valid when this loops happens.

> If you hit No on the "download db dialog" you'll have a open and running Cockatrice with no cards, but after a few long seconds the old "Welcome, first run" msg appears and explains you that you need to manage your sets first (so it somehow still detects that you now have some cards, even the reloading didn't work!),

This is not expected. Either the database load is successful, and the "new sets check" starts, or the database load fails and the "download db dialog" pops up. I need to check what's going on here.

> you can just hit ok (actually hitting x will act the same as hitting ok - so you can't avoid it which is good and right here) and the edit sets window opens with all sets enabled by default.

If cockatrice loaded the sets list, in reality the card database was loaded successfully.. strange

> Pressing ok forces Cockatrice to reload the database

Hum.. it doesn't really reloads the database, it just updates the filters on the deck_editor that includes/excludes cards depending on the enabled/disabled sets. So this means the card database already loaded the cards, but no set was enabled.

> if you hit x or cancel Cockatrice shows up no cards,

I suppose this confirms that cards were loaded, but no set was enabled.

> if you restart Cockatrice though all sets are enabled and show up - also a nice default if someone just clicks trough.

Ok

> This needs some tweaks! (checks are mistimed? refresh is triggered but leads to a repeating "download db" msg instead of a reload and the "welcome, first run" msg to appear...)

I suppose that Cockatrice loaded the card database, but somehow got confused by the set list.. maybe it's not clearing / reloading /enabling the sets list correctly. I'll try to check this out.

ctrlaltca commented:
Ok, the problem reported by @tooomm should be fixed now.
I've also added a small fix for #1546 and squashed commits down.

radioactivehazards commented:
So does this work now?? because i still cannot see my SOI pictures for the Tokens I have made.

ctrlaltca commented:
> So does this work now??

It works for me, if you want to test it out you can download an experimental build at https://ci.appveyor.com/project/Daenyth/cockatrice-8k7y5/build/0.0.1-branch-master-build-124 : scroll to the bottom, choose the first line (Environment: vc_arch=amd64) for 64bit or the second line (Environment: vc_arch=amd64_x86) for 32bit, the click on "artifacts" to get the link to the installer.

radioactivehazards commented:
are you putting the fix into the next "daily/weekly" build?  or whatever it is. not sure i like the idea of "experimental" build

ctrlaltca commented:
As soon as testers will :+1: this, it will probably be merged and end up in an official build, yes.

radioactivehazards commented:
Not only did that NOT work. it killed my Custom card backs. and loaded the decks weird as now they have an "unknown" card area and the new SOI cards are not an unsorted pile on top.

radioactivehazards commented:
omg.....it killed all my themes......Great!!! thx guys.

radioactivehazards commented:
yeah Every piece of my Custom theme is now gone............so yeah. i guess Cockatrice is Forever broken.

ctrlaltca commented:
Please keep it into the topic of this PR, this is not your personal shoutbox.
If you had problem testing the experimental build, please report them in detail so that we can understand what went wrong and how to fix it. If you ran the experimental build on a pc with valuable data and no backup, i'm sorry for you.

radioactivehazards commented:
nah its fine i feel for your False attempts at help and am now screwed. and I have no idea on how to fix cockatrice because you guys have messed it up so badly now. I Reinstalled the Version i had before this conversation and as Always whenever i take your advice I get screwed. So yeah I Quit.

ctrlaltca commented:
> your False attempts at help
> you guys have messed it up so badly now
> whenever i take your advice I get screwed

That's exactly what better motivates us developers to work for free at projects on our spare time, thank you.

radioactivehazards commented:
The project in Question was fine.......needed no fixing......you decided to
"fix" things and actually made things worse. I apologize a little but t
seems the more i try to address the Problems you created the more problems
I get and the more people get MAd at me for trying to help. You tell me to
suggest things but i either get yelled at each time or i have to jump
through so many Damn hoops its just annoying.

On Thu, Mar 17, 2016 at 5:00 PM, ctrlaltca <notifications@github.com> wrote:

> your False attempts at help
> you guys have messed it up so badly now
> whenever i take your advice I get screwed
>
> That's exactly what better motivates us developers to work for free at
> projects on our spare time, thank you.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/Cockatrice/Cockatrice/pull/1859#issuecomment-198080699>
>

Daenyth commented:
@radioactivehazards I'm not sure why your settings got removed; I'm sorry they did. This was pretty clearly marked as an experimental build though, and not a full release. There's always some element of risk with that. People are upset because you were using an experimental build and then getting upset when it didn't work correctly, saying things like "false help". 

Now, for the theme stuff - it should be possible to reconfigure it. I doubt it's broken forever; in fact given that this change affects the way card images are loaded, it's much more likely that *image loading* is broken, not your installation.

If you reinstall a normal release, does it work again?

radioactivehazards commented:
No. i have to Rebuild my theme From scratch and thats a pain as this is
probably the 56th time I have done so. When i quote False help i mean the
things that worked fine before they tinkered and broke them like the
Original theme set up. The Original sound set up. Not the Custom tokens and
ways to proxy cards are getting "changed" so everything that works is being
systematically Destroyed. also i hear that they are killing the 32 bit set
up for cock soon so its kinda obvious what is being done.

On Thu, Mar 17, 2016 at 5:16 PM, Gavin Bisesi <notifications@github.com>
wrote:

> @radioactivehazards <https://github.com/radioactivehazards> I'm not sure
> why your settings got removed; I'm sorry they did. This was pretty clearly
> marked as an experimental build though, and not a full release. There's
> always some element of risk with that. People are upset because you were
> using an experimental build and then getting upset when it didn't work
> correctly, saying things like "false help".
>
> Now, for the theme stuff - it should be possible to reconfigure it. I
> doubt it's broken forever; in fact given that this change affects the way
> card images are loaded, it's much more likely that *image loading* is
> broken, not your installation.
>
> If you reinstall a normal release, does it work again?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/Cockatrice/Cockatrice/pull/1859#issuecomment-198085636>
>

leematos commented:
@radioactivehazards, what platform are you on? I'm happy to set aside some time (30 minutes or so) tomorrow/early next week to personally help you with this if you are interested. This is outside of the scope of this PR so we'll need to connect via some other way. You can find my email on my personal site. Reach out. (http://leematos.com/)

Daenyth commented:
@radioactivehazards I'm not sure what "they" you're talking about; it's been just the few of us as main contributors for many years...

You shouldn't need to rebuild your theme - where was it being stored? In your AppData folder? Custom tokens being broken is a bug which we should be fixing for the next release. This change doesn't break anything, it's a change under the hood to keep the same behavior but be easier to tweak later and perform faster.

I don't believe we have plans to remove the 32bit setup either, at least not any time soon.

radioactivehazards commented:
Lee, Thankyou but you sir are too late. it seems every feature i enjoyed is
being wrecked 1 by 1 so ill just stop enjoying things. they or whoever
works on the service killed themes, sounds and proxy methods. I will admit
the related card feature was nice and that hopefully will not be wrecked
any time soon.

On Thu, Mar 17, 2016 at 5:51 PM, Gavin Bisesi <notifications@github.com>
wrote:

> @radioactivehazards <https://github.com/radioactivehazards> I'm not sure
> what "they" you're talking about; it's been just the few of us as main
> contributors for many years...
>
> You shouldn't need to rebuild your theme - where was it being stored? In
> your AppData folder? Custom tokens being broken is a bug which we should be
> fixing for the next release. This change doesn't break anything, it's a
> change under the hood to keep the same behavior but be easier to tweak
> later and perform faster.
>
> I don't believe we have plans to remove the 32bit setup either, at least
> not any time soon.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/Cockatrice/Cockatrice/pull/1859#issuecomment-198099560>
>

Daenyth commented:
@radioactivehazards I'm locking this thread so we can keep it on topic.

ctrlaltca commented:
PR Unlocked, OT comments deleted.

Daenyth commented:
Because this change is so big, while we are testing it, I would prefer it if you do not rebase so that we can review changes commit-by-commit as we get feedback.

When feedback is done we can rebase+squash as normal.

Daenyth commented:
[Current exe link](https://ci.appveyor.com/api/buildjobs/ltt7bn04dhj2tqcm/artifacts/Cockatrice-0.0.1~git20160321.6601ed2-x86_64_qt5.exe)

tredzko commented:
Initial problems (I'll be able to test more when I get home later):
- Dragging from the top of the library to the battlefield causes an exception when the top card is not revealed.
- I believe zooming should happen on the field with the different phases and the library staying a constant size, right? If that's the case, then currently this version zooms everything out.

ctrlaltca commented:
@tredzko thank you for the feedback.

> Dragging from the top of the library to the battlefield causes an exception when the top card is not revealed.

You found a long-time bug, that previously caused only a minor problem but now is amplified by this PR. I've detailed the bug at #1882. In the meanwhile, i've patched the problem in this PR.


> I believe zooming should happen on the field with the different phases and the library staying a constant size, right? If that's the case, then currently this version zooms everything out.

As far as i remember the whole gamearea zooms out, including the "deck/graveyard/exile" and the "phases" zones. If that behavior has changed, i don't think it's due to this PR.

Daenyth commented:
Once appveyor builds correctly, let me know and I'll update the download link for people

Daenyth commented:
@tredzko I updated my link with the latest version including crash fix. https://github.com/Cockatrice/Cockatrice/pull/1859#issuecomment-199538393
